File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
class Solution:
    def reverse(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        curr = head
        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node
        return prev

    def reorderList(self, head: Optional[ListNode]) -> None:
        if not head or not head.next:
            return

        # 1. Find the middle of the list
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # 2. Split the list and reverse the second half
        second = self.reverse(slow.next)
        slow.next = None  # Sever the connection
        
        # 3. Merge the two halves
        first = head
        while second:
            # Save next pointers
            tmp1, tmp2 = first.next, second.next
            
            # Re-wire pointers
            first.next = second
            second.next = tmp1
            
            # Move pointers forward
            first = tmp1
            second = tmp2

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(!head){return NULL;}
        Node* curr=head;
        while(curr){
            Node* nextNode = curr->next;
            curr->next = new Node(curr->val);
            curr->next->next=nextNode;
            curr=curr->next->next;
        }
        curr=head;
        while(curr){
            if(curr->random){curr->next->random = curr->random->next;}
            curr=curr->next->next;
        }
        curr=head;
        Node* NewHead = head->next;
        while(curr){
            Node* temp=curr->next;
            curr->next=temp->next;
            if(temp->next){
                temp->next=temp->next->next;}
            curr=curr->next;
        }
        return NewHead;
    }
};
